File: bot.py
from lib2to3.fixes.fix_input import context
import logging
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Application, CommandHandler, CallbackQueryHandler, ContextTypes, MessageHandler, filters
import requests
from datetime import datetime

# Load API keys from environment variables
TMDB_API_KEY = '69084ded6889a849708077681bd5dd7f'
BOT_TOKEN = '7501751569:AAETZdiv2Er_l8YVSciALtEzvY5ic12figo'
BASE_URL = 'https://image.tmdb.org/t/p/w300'

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger()

async def start(update: Update, context):
    await update.message.reply_text("Welcome to the Movie Explorer Bot! 🎬\nType /search [movie name] to find movies or TV shows.")

async def search(update: Update, context):
    query = ' '.join(context.args)
    if not query:
        await update.message.reply_text("Please provide a search query. Example: /search Inception")
        return
    await handle_search(update.message.chat_id, query, context)

async def handle_search(chat_id, query, context, page=1):
    apiUrl = f'https://api.themoviedb.org/3/search/multi?api_key={TMDB_API_KEY}&query={query}&page={page}'
    response = requests.get(apiUrl)
    data = response.json()
    total_pages = data.get('total_pages', 0)

    if data.get('results') and len(data['results']) > 0:
        results = data['results'][:5]
        buttons = []
        for result in results:
            title = result.get('title') or result.get('name')  # Safely getting title or name
            media_type = result.get('media_type', 'movie')    # Default to 'movie' if not specified
            if title:  # Ensure there's a title or name to create a button
                buttons.append([InlineKeyboardButton(text=title, callback_data=f"detail_{result['id']}_{media_type}")])

        navigation_buttons = []
        if page > 1:
            navigation_buttons.append(InlineKeyboardButton(text='← Prev', callback_data=f"change_{query}_{page - 1}"))
        navigation_buttons.append(InlineKeyboardButton(text=f"{page}/{total_pages}", callback_data='noop'))
        if page < total_pages:
            navigation_buttons.append(InlineKeyboardButton(text='Next →', callback_data=f"change_{query}_{page + 1}"))

        if navigation_buttons:
            buttons.append(navigation_buttons)

        reply_markup = InlineKeyboardMarkup(buttons)
        await context.bot.send_message(chat_id, 'Select a title to get more details:', reply_markup=reply_markup)
    else:
        await context.bot.send_message(chat_id, f'No results found for "{query}".')

async def handle_details(chat_id, id, type, context):
    apiUrl = f'https://api.themoviedb.org/3/{type}/{id}?api_key={TMDB_API_KEY}&append_to_response=images'
    response = requests.get(apiUrl)
    data = response.json()

    title = data.get('title') or data.get('name')  # Safely getting title or name
    release_date = data.get('release_date') or data.get('first_air_date')
    vote_average = data.get('vote_average', 'N/A')
    overview = data.get('overview', 'No description available.')
    posterUrl = f"{BASE_URL}{data.get('backdrop_path')}" if data.get('backdrop_path') else None

    release_datetime = datetime.strptime(release_date, '%Y-%m-%d') if release_date else None
    today = datetime.today()

    if release_datetime and release_datetime < today:
        message_text = f"The movie or TV show has already been released on {release_date}."
        reply_markup = None  # Ensure reply_markup is always defined
    else:
        message_text = (
            f"*{title}*\n"
            f"📅 Release Date: {release_date}\n"
            f"⭐ Rating: {vote_average}/10\n"
            f"{overview}"
        )
        if release_datetime:
            buttons = [[InlineKeyboardButton(text="🔔 Remind Me", callback_data=f"remind_{chat_id}_{id}_{release_date}")]]
            reply_markup = InlineKeyboardMarkup(buttons)
        else:
            reply_markup = None  # Define reply_markup even if there's no future release date

    if posterUrl:
        await context.bot.send_message(chat_id, f'<a href="{posterUrl}">&#8205;</a>{message_text}', parse_mode='HTML', reply_markup=reply_markup)
    else:
        await context.bot.send_message(chat_id, message_text, parse_mode='Markdown', reply_markup=reply_markup)

async def callback_handler(update: Update, context):
    query = update.callback_query
    await query.answer()
    data = query.data.split('_')
    action = data[0]
    param1 = data[1]
    if action == 'detail':
        await handle_details(query.message.chat_id, param1, data[2], context)
    elif action == 'change':
        await handle_search(query.message.chat_id, param1, context, int(data[2]))

def main():
    app = Application.builder().token(BOT_TOKEN).build()

    app.add_handler(CommandHandler("start", start))
    app.add_handler(CommandHandler("search", search))
    app.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, search))
    app.add_handler(CallbackQueryHandler(callback_handler))

    app.run_polling()

if __name__ == '__main__':
    main()

File: bot.js
// Constants for API keys, should be loaded from environment variables
const TMDB_API_KEY = '69084ded6889a849708077681bd5dd7f';
const BOT_TOKEN = '8153935470:AAETIPoq3zdQg2kDFvmxSgVQ9Wz_1R8ophU';
const BASE_URL = 'https://media.themoviedb.org/t/p/w300_and_h450_bestv2';

// Function to send messages to the Telegram chat
async function sendMessage(chatId, text, options = {}) {
    const url = `https://api.telegram.org/bot${BOT_TOKEN}/sendMessage`;
    const body = { chat_id: chatId, text, ...options };

    const response = await fetch(url, {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify(body),
    });

    return await response.json();
}

// Handle incoming messages
async function handleMessage(message) {
    const chatId = message.chat.id;
    const text = message.text;

    if (text === '/start') {
        const welcomeMessage = "Welcome to the Movie Explorer Bot! 🎬\nType /search [movie name] to find movies or TV shows.";
        await sendMessage(chatId, welcomeMessage);
    } else if (text.startsWith('/search')) {
        const query = text.split(' ').slice(1).join(' ');
        if (!query) {
            await sendMessage(chatId, "Please provide a search query. Example: /search Inception");
            return;
        }
        await handleSearch(chatId, query);
    }
}

// Function to search for media and display results with pagination
async function handleSearch(chatId, query, page = 1) {
    const apiUrl = `https://api.themoviedb.org/3/search/multi?api_key=${TMDB_API_KEY}&query=${encodeURIComponent(query)}&page=${page}`;
    try {
        const response = await fetch(apiUrl);
        const data = await response.json();
        const total_pages = data.total_pages;

        if (data.results && data.results.length) {
            const results = data.results.slice(0, 5);
            const inlineKeyboard = {
                inline_keyboard: [
                    ...results.map(result => [{
                        text: result.title || result.name,
                        callback_data: `detail_${result.id}_${result.media_type}`
                    }]),
                    [
                        {text: '← Prev', callback_data: `change_${query}_${Math.max(page - 1, 1)}`},
                        {text: `${page}/${total_pages}`, callback_data: 'noop'},
                        {text: 'Next →', callback_data: `change_${query}_${Math.min(page + 1, total_pages)}`}
                    ].filter(btn => btn.callback_data.includes(String(page)) || btn.text === 'noop')
                ]
            };
            await sendMessage(chatId, 'Select a title to get more details:', {reply_markup: inlineKeyboard});
        } else {
            await sendMessage(chatId, `No results found for "${query}".`);
        }
    } catch (error) {
        await sendMessage(chatId, `Error searching for "${query}": ${error.message}`);
    }
}

// Function to display detailed information with poster
async function handleDetails(chatId, id, type) {
    const apiUrl = `https://api.themoviedb.org/3/${type}/${id}?api_key=${TMDB_API_KEY}&append_to_response=images`;
    try {
        const response = await fetch(apiUrl);
        const data = await response.json();
        const details = `*${data.title || data.name}*\n_Release Date_: ${data.release_date || data.first_air_date}\n_Rating_: ${data.vote_average}\n\n${data.overview}`;
        const posterUrl = data.backdrop_path ? `${BASE_URL}${data.backdrop_path}` : null;

        if (posterUrl) {
            await sendMessage(chatId, `<a href="${posterUrl}">&#8205;</a>${details}`, { parse_mode: 'HTML' });
        } else {
            await sendMessage(chatId, details, { parse_mode: 'Markdown' });
        }
    } catch (error) {
        await sendMessage(chatId, `Failed to fetch details: ${error.message}`);
    }
}

// Function to handle callback queries from inline keyboards
async function handleCallbackQuery(callbackQuery) {
    const chatId = callbackQuery.message.chat.id;
    const data = callbackQuery.data;
    const [action, param1, param2] = data.split('_');

    if (action === 'detail') {
        await handleDetails(chatId, param1, param2);
    } else if (action === 'change') {
        if (param2 > 0) {
            await handleSearch(chatId, param1, parseInt(param2));
        }
    }
}

// Main event listener for incoming HTTP requests
addEventListener('fetch', event => {
    event.respondWith(handleRequest(event.request));
});

// Handle incoming HTTP requests
async function handleRequest(request) {
    if (request.url.endsWith('/webhook')) {
        const update = await request.json();
        if (update.message) {
            await handleMessage(update.message);
        } else if (update.inline_query) {
            // Implement if required
        } else if (update.callback_query) {
            await handleCallbackQuery(update.callback_query);
        }
        return new Response('OK', { status: 200 });
    }
    return new Response('Not Found', { status: 404 });
        }
